### ISSUE TITLE
Duplicated messages in Kafka during delayed publishing large messges

### ISSUE THREAD

**Issue by PoteRii** [dotnetcore/CAP#1429]
We are seeing incorrect behavior of publishing delayed messages in production so I replicated the issue on local with quite simple scenario.

```c#
using DotNetCore.CAP;
using KafkaTest;

WebApplicationBuilder builder = WebApplication.CreateBuilder(args);

builder.Services.AddTransient<ISubscriberService, SubscriberService>();

builder.Services.AddCap(options =>
{
    options.UseDashboard();
    options.UsePostgreSql("Server=localhost;User ID=postgres;Password=1234;Database=CapTest");
    options.UseKafka(kafkaOptions => { kafkaOptions.Servers = "localhost:9092"; });

    options.UseStorageLock = true;
});

WebApplication app = builder.Build();

app.MapGet("/test", async (ICapPublisher eventBus) =>
{
    for (int i = 0; i < 10000; i++)
    {
        await eventBus.PublishDelayAsync(TimeSpan.FromMinutes(5), "cap-test", Guid.NewGuid().ToString().ToLower());
    }
});

app.Run();
```

After startup of this project 10k messages should be published in Kafka in 5 minutes, but the result is that it`s always more than 10k (sometimes 13k, sometimes more). Below you can find the info when 22862 messages were produced.

GUID duplicated at least 3 times
![Screenshot 2023-11-13 111340](https://github.com/dotnetcore/CAP/assets/43849591/5524067e-966e-4173-9c3a-8c4319e5d118)

All the headers are exactly the same
![Screenshot 2023-11-13 111354](https://github.com/dotnetcore/CAP/assets/43849591/dd522121-3fb1-4d5e-96e5-89da9a5adf9d)
![Screenshot 2023-11-13 111406](https://github.com/dotnetcore/CAP/assets/43849591/c0aed588-b6e7-498d-a8a5-34c0c81f2283)

No messages have Retries > 0
![Screenshot 2023-11-13 111503](https://github.com/dotnetcore/CAP/assets/43849591/473bc82c-7c02-4458-a739-d1a40358cf8a)

22862 produced to the topic (confirmed by making separate consumer)
![Screenshot 2023-11-13 111521](https://github.com/dotnetcore/CAP/assets/43849591/94997c7d-4811-4c2f-8786-744cc4e9623d)

All of the below messages are duplicates
![Screenshot 2023-11-13 114110](https://github.com/dotnetcore/CAP/assets/43849591/e3ae1ac7-cd64-4a77-a364-cafdd3fd3b48)

Packages used:
DotNetCore.CAP.Dashboard Version 7.2.2
DotNetCore.CAP.Kafka Version 7.2.2
DotNetCore.CAP.PostgreSql Version 7.2.2

**Comment by PoteRii** [comment]
UPDATE:
MessageDelayedProcessor.ProcessDelayedAsync starts running and picks up messages from db while previous run has not finished processing 10k messages, hence the duplication in kafka appears

**Comment by PoteRii** [comment]
So, what I think happens is
- Delayed or Queued messages are taken from published table
- There are plenty of them so processing takes time
- meanwhile ProcessDelayedAsync is fired again (it`s fired every one minute)
- Messages that are already in PriorityQueue are picked up again and published again

**Comment by PoteRii** [comment]
UPDATE:
this piece of code is executed, so this means that assumption in my previous comment is correct.
![Screenshot 2023-11-13 172614](https://github.com/dotnetcore/CAP/assets/43849591/f91763b9-10ed-4a16-adcb-9660b88b5bb3)


"FOR UPDATE SKIP LOCKED" in `ScheduleMessagesOfDelayedAsync` does not do much in respective of avoiding this duplication, because transaction is opened and committed as soon as messages are put in priority queue, so it makes those messages available for the next iteration of `MessageDelayedProcessor.ProcessDelayedAsync`

**Comment by yang-xiaodong** [comment]
Hi @PoteRii ,

Thank you for your feedback. I took a quick look and as you said when there are a large number of delayed messages(same time trigger) and the delayed memory queue are not completely processed, duplication will occur when new messages are picked up by `MessageDelayedProcessor.ProcessDelayedAsync`. When doing this feature, the situation of a large number of delayed messages was not considered.

The reason we are not using StorageLock in publish delay messages is that some people don't want to add extra lock table and want to use send delay message feature, and we used "FOR UPDATE SKIP LOCKED"  can prevent multiple instances from getting the same message, because before placing it in the PriorityQueue, the message status will change to Queued before the transaction commited.

**Comment by yang-xiaodong** [comment]
UPDATE:
one question, Before the transaction is committed, the FOR UPDATE query locks a batch of data, and we update the status to Queued within the scope of the transaction.  Why is this data still being queried and enqueue?

**Comment by PoteRii** [comment]
@yang-xiaodong  Thank you for the reply. `ChangePublishStateAsync` is called from `EnqueueToScheduler` which is called from the loop over 10k messages. So I guess this loop takes more than 1 min, so that`s why the messages that has not yet been updated are picked up by next process.

**Comment by PoteRii** [comment]
I have added some logging in `ScheduleMessagesOfDelayedAsync` and changed `EnqueueToScheduler` method to return `true` if message was added to the priority queue, than I am checking that boolean result in `ProcessDelayedAsync.ScheduleTask` and increasing count of queued messages (3rd line in the logs).

The logs I see are below
![image](https://github.com/dotnetcore/CAP/assets/43849591/8ec0e73d-1c76-43e4-a624-5d1f1abe8e14)

This confirms the situation happening in the previous comment.
P.S. I`ve stopped app in the end.

**Comment by yang-xiaodong** [comment]
Hi @PoteRii,

Thanks for your more information, I need some time to think about this problem, looks a bit difficult and I am busy recently.

Can you test whether simply changing 

https://github.com/dotnetcore/CAP/blob/7ba52a8defa73f3866001dea6dd6af434574c3d2/src/DotNetCore.CAP/Processor/IDispatcher.Default.cs#L115

to

```
await EnqueueToPublish(_schedulerQueue.Dequeue()).ConfigureAwait(false);
```
can send 10k data in 1 minute?

**Comment by PoteRii** [comment]
@yang-xiaodong 
As a matter of fact, my system froze a bit, but EnqueueToPublish was able to handle 10k messages almost in 1 min. Publish TPS was around 10x higher (around 500-600 compared to 50-60 for SendAsync).

Although I think this is a logical issue rather than performance, because I can have X number of messages, that can`t be handled in 1 min by EnqueueToPublish

**Comment by yang-xiaodong** [comment]
@PoteRii 

I have an idea. In the processor for picking up delayed messages (ScheduleMessagesOfDelayedAsync), we remove the filter condition for messages with the status 'Queued'. Instead, we only pick up messages with the 'Queued' status and place them into the priority queue during the first application startup.

In scenarios with multiple instances, when UseStorageLock is enabled, we can use database locks to prevent different instances from picking up the same 'Queued' messages at startup. In this case, we need to consider a situation where the instance that acquired the database lock stops or crashes within the TTL (Time To Live) period. In such cases, these messages will need to wait to be triggered until the next instance startup after the TTL has expired.

**Comment by PoteRii** [comment]
@yang-xiaodong If I get this right, the messages can get stuck in `Queued` status only in case of ungraceful shutdown. If that`s the case than it makes sense to process those stuck messages only during app startup. TTL is `FailedRetryInterval` + 10 seconds right? So that will give the app 70 sec (by default) to process all the `Queued` messages. Are my assumptions correct?

**Comment by yang-xiaodong** [comment]
@PoteRii 

The main issue is not whether all `Queued` messages can be all processed within 70 seconds, but rather if the application crashes or restarts again within the 70 seconds, it will not be possible to read the Queued messages because the TTL has not yet expired. This is especially the case in Kubernetes, where the host name (instance id) changes after Pod restarts, and at this time the TTL has not exceeded 70 seconds, thus making it impossible to acquire the lock.

**Comment by yang-xiaodong** [comment]
@PoteRii 

Following the above idea, there is another problem that cannot be solved: if there is a very large number of messages with the status 'Queued', we cannot load them all into memory at once. However, if we fetch messages in batches, it's impossible to know whether the previous batch has finished processing in order to decide whether to RenewLock.

One idea is to wait for a fixed duration and repeatedly check for messages still in the 'Queued' state. This means that within the TTL lifecycle, only a fixed number of messages can be processed. For example, if we process 500 messages per batch and there are 10K messages with a TTL of 60 seconds, it would take 20 minutes to process all messages. Moreover, each batch of 500 messages must be processed within 60 seconds, or else the issue of duplicates would still persist.

**Comment by PoteRii** [comment]
@yang-xiaodong would something like this work? 
Processor tries to acquire to lock
   Not acquired
      Try again after lock TTL expires
   Acquired
      Fetch queued messages one by one (or by batches)
     Send messages in a sync manner
     Renew lock once in a while before TTL expires but there are still messages to be processed

**Comment by demorgi** [comment]
@PoteRii Consider a few additional things, producer guarantess at-least-once delivery same as broker does, so it's on a consumer side to make sure idempotent processing.
@yang-xiaodong maybe it would make sense just to make batch size and time window to be configurable, as it won't be possible to know thoughtput of a client system, as at the moment there is no reliable way to know where there is a running sending process.

**Comment by PoteRii** [comment]
@demorgi It is natural to have duplications once in a while caused by a "glitch", but this case is quite different. One run can produce 25k messages insted of 10k. With a duplication rate for a message as high as 6-8 times.

**Comment by yang-xiaodong** [comment]
@PoteRii 
> Send messages in a sync manner

Delayed messages cannot be sent synchronously, as these messages may not have reached the actual time for sending. This messages have a time window of 1 minute.

**Comment by yang-xiaodong** [comment]
One solution is the messages support send completion callback. Set a callback for the last message of each batch (if they are delayed messages, they need to be sorted by send time), and fetch the next batch when the callback is triggered. Before this, the thread is paused, and the callback will resume the thread to continue running.

**Comment by PoteRii** [comment]
@yang-xiaodong good idea

**Comment by yang-xiaodong** [comment]
@PoteRii 
I've encountered another issue. Suppose I have five instances that have been running for a while, and one of them gets stopped. Instead of restarting it, I decide to continue running the remaining four instances. In this scenario, the stopped instance might have unprocessed Queued messages, which will remain unhandled indefinitely until the next time an instance is started.

**Comment by PoteRii** [comment]
@yang-xiaodong what about showing queued messages (that have been queued for quite long) in dashboard?

**Comment by yang-xiaodong** [comment]
That's not a good idea. If the Queued messages use the approach we discussed, then the messages in the published and received tables with the status 'scheduled' will be handled using the same approach.

**Comment by PoteRii** [comment]
@yang-xiaodong Let's assume there is a process that starts on app startup. It acquires lock (or retries until it can acquire it), fetches queued messages and processes all of them in batches. What keeps me from running this process every X minutes? So basically this is similar to adding locking mechanism to already existing processor...

**Comment by yang-xiaodong** [comment]
@PoteRii 

> What keeps me from running this process every X minutes?

Because if every X minutes, this might easy picked up messages whose status has just been changed to Queued within a 1-minute window (Status from delay to queued not lock), when in fact these messages are already put into the priority queue.

**Comment by PoteRii** [comment]
@yang-xiaodong I think the whole queuing concept has to be revised... Meanwhile the easiest short term solution would be to make `OneMinutesAgo` query parameter configurable, to at least give fetched delayed messages more than 1 minute to be processed.

**Comment by yang-xiaodong** [comment]
As a short term solution, see commit https://github.com/dotnetcore/CAP/commit/aa4c54041a8a990838523d2b68241fe56658a702 , you need override `QueuedMessageFetchTime`

**Comment by yang-xiaodong** [comment]
Fixed in version `7.2.3-preview-217174309`

**Comment by PoteRii** [comment]
@yang-xiaodong tested and can confirm this works as intended. Thank you.
